kubectl/upload/main.go

```diff
@@ -0,0 +1,212 @@
+	}
+
+	if cfg.Verbose {
+		fmt.Println("packaging the directory into a tarball")
+	}
+
+	tarPath := "/tmp/archive.tar.gz"
+	err := tarGz(cfg.Path, tarPath)
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tarPath)
+
+	checksum, err := calculateMD5(tarPath)
+	if err != nil {
+		return err
+	}
+
+	cfg.Resource.Md5Checksum = checksum
+	data, err := hex.DecodeString(checksum)
+	if err != nil {
+		return err
+	}
+	cfg.Resource.EncodedMd5 = base64.StdEncoding.EncodeToString(data)
+
+	customResource := createCustomResource(cfg)
+
+	gvr := schema.GroupVersionResource{
+		Group:    "substratus.ai",
+		Version:  "v1",
+		Resource: strings.ToLower(cfg.Resource.Kind) + "s", // plural is needed here
+	}
+
+	result, err := client.CreateResource(gvr, cfg.Resource.Namespace, customResource, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create custom resource: %w", err)
+	}
+
+	watcher, err := client.WatchResource(gvr, cfg.Resource.Namespace, metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("metadata.name=%s", customResource.GetName()),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to watch resource %q: %w", customResource.GetName(), err)
+	}
+	if cfg.Verbose {
+		fmt.Printf("Created custom resource %q.\n", result.GetName())
+	}
+
+	// Monitor the watcher channel
+	for event := range watcher.ResultChan() {
+		switch event.Type {
+		case watch.Added, watch.Modified:
+			err = handleWatchEvent(event, cfg, tarPath)
+			if err != nil {
+				return err
+			}
+		case watch.Error:
+			return errors.New("encountered a watch error")
+		case watch.Deleted:
+			return errors.New("the custom resource was deleted")
+		default:
+			return errors.New("unhandled event type")
+		}
+	}
+
+	if cfg.Verbose {
+		fmt.Printf("Upload is complete. Waiting for the build job to complete.\n")
+	}
+	// TODO(bjb): should we put a watcher also on name-kind-container-builder or does the utility stop here?
+	// spin.Start()
+	// spin.Suffix = " Waiting for the build job to complete..."
+	// spin.Stop()
+	// TODO(bjb): if it's a notebook, wait for it to be ready, then open it in a browser
+
+	return nil
+}
+
+func createCustomResource(cfg Config) *unstructured.Unstructured {
+	return &unstructured.Unstructured{Object: map[string]interface{}{
+		"apiVersion": "substratus.ai/v1",
+		"kind":       cfg.Resource.Kind,
+		"metadata": map[string]interface{}{
+			"name": cfg.Resource.Name,
+		},
+		"spec": map[string]interface{}{
+			"image": map[string]interface{}{
+				"upload": map[string]interface{}{
+					"md5checksum": cfg.Resource.Md5Checksum,
+				},
+			},
+		},
+	}}
+}
+
+func handleWatchEvent(event watch.Event, cfg Config, tarPath string) error {
+	updatedResource := event.Object.(*unstructured.Unstructured)
+
+	// Retrieve the value of .status.upload.uploadURL
+	status, ok, err := unstructured.NestedMap(updatedResource.Object, "status", "upload")
+	if err != nil || !ok {
+		return nil
```

Comment:
Should this be `return err`?

Reply:
Actually: `if err !=nil { return err`

... And: `if !ok { return errors.New("no upload status")`
?

Reply:
Yeah, I was anticipating this watch will not have a `status.upload` immediately so I thought it was the right move to `return nil` but I'll spend some time here figuring out the correct behavior.

Reply:
Ahh, ok that makes sense